File: Practico3/Fibonacci/Fibonacci.hs
-- F0 = 0, F1 = 1, F2 = 1, F3 = 2, F4 = 3, F5 = 5, F6 = 8

-- Fibonacci sequence (inefficient version)
fib :: Int -> Int
fib 0 = 0
fib 1 = 1
fib n = (fib (n-1)) + (fib (n-2))

-- Memoization version of Fibonacci sequence
fibo :: Int -> Int
fibo x = lookUp x (memoFibo x [])

memoFibo :: Int -> [(Int, Int)] -> [(Int,Int)]
memoFibo x cache | defined cache x = cache
                 | x<0 = error "undefined"
                 | x<=1 = store cache x 1
                 | otherwise = store cacheSnd x (y1+y2)
    where cacheFst = memoFibo (x-1) cache
          cacheSnd = memoFibo (x-2) cacheFst
          y1 = lookUp (x-1) cacheSnd
          y2 = lookUp (x-2) cacheSnd

lookUp :: Int -> [(Int, Int)] -> Int
lookUp x [] = error "undefined"
lookUp x1 ((x2,y):ps) | x1==x2 = y
                      | x1/=x2 = lookUp x1 ps

defined :: [(Int,Int)] -> Int -> Bool
defined ps x = elem x (map fst ps)

store :: [(Int, Int)] -> Int -> Int -> [(Int,Int)]
store ps x y | defined ps x = ps
             | otherwise = (x,y):ps

File: Practico3/Fibonacci/Fibonacci.java
import java.util.HashMap;
import java.util.Map;

/**
 * Algoritmo basado en programacion dinamica que calcula el n-esimo numero de Fibonacci
 * -- F0 = 0, F1 = 1, F2 = 1, F3 = 2, F4 = 3, F5 = 5, F6 = 8
 */
public class Fibonacci {
    
    public static void main(String[] args){
        System.out.println(memoFib(6));
    }

    // Version ineficiente recursiva
    public static int fibRec(int n){
        if (n < 0) throw new IllegalArgumentException("undefined for negative integers");
        if (n == 0) return 0;
        if (n == 1) return 1;

        return fibRec(n-1) + fibRec(n-2);
    }
    
    // Version eficiente con PD 
    public static int fib(int n){
        if (n == 0) return 0;

        int fibNum1 = 0; 
        int fibNum2 = 1;
        int aux;

        for (int i = 1; i < n; i++){
            aux = fibNum1 + fibNum2;
            fibNum1 = fibNum2;
            fibNum2 = aux;
        }

        return fibNum2;
    }

    // Structure that represents the cache
    private static Map<Integer, Integer> cache = new HashMap<Integer, Integer>();

    // Version eficiente con Memoization
    public static int memoFib(int n){
        if (n < 0) throw new IllegalArgumentException("undefined for negative integers");

        if (!cache.containsKey(n)){
            if (n == 0){
                cache.put(n,0);
            }
            if (n == 1){
                cache.put(n,1);
            }
            if (n > 1){
                cache.put(n, memoFib(n-1) + memoFib(n-2));
            }
        }
        return cache.get(n);
    }
}
